src/client/opamArg.ml

```diff
@@ -782,62 +774,414 @@ let opamlist_columns =
+    (validity * 'a * string list * string) list -> 'a Term.t
+
+  val mk_vflag_all:
+    cli:OpamCLIVersion.t -> ?section:string -> ?default:'a list ->
+    (validity * 'a * string list * string) list -> 'a list Term.t
+
+  val mk_tristate_opt:
+    cli:OpamCLIVersion.t -> validity -> ?section:string -> string list ->
+    string -> string -> [> `Always | `Auto | `Never ] option Term.t
+
+  type 'a subcommand = validity * string * 'a * string list * string
+  type 'a subcommands = 'a subcommand list
 
-let mk_subcommands commands =
-  mk_subcommands_aux Arg.enum commands
+  val mk_subcommands:
+    cli:OpamCLIVersion.t -> 'a subcommands ->
+    'a option Term.t * string list Term.t
 
-let mk_subcommands_with_default commands =
-  mk_subcommands_aux enum_with_default commands
+  val mk_subcommands_with_default:
+    cli:OpamCLIVersion.t -> 'a default subcommands ->
+    'a option Term.t * string list Term.t
+
+  val bad_subcommand:
+    cli:OpamCLIVersion.t -> 'a default subcommands ->
+    (string * 'a option * string list) -> 'b Term.ret
+
+  val mk_subdoc :
+    cli:OpamCLIVersion.t -> ?defaults:(string * string) list ->
+    'a subcommands -> Manpage.block list
+
+end = struct
+
+  let cli2_0 = OpamCLIVersion.of_string "2.0"
+  let cli2_1 = OpamCLIVersion.of_string "2.1"
+
+  type 'b validity_and_content = {
+    valid: OpamCLIVersion.t;
+    removed: (OpamCLIVersion.t * string option) option;
+    content: 'b;
+  }
+
+
+  type 'a contented_validity =
+    [ `Valid of 'a | `Removed of 'a ] validity_and_content
+
+  type validity = [ `Nothing ] validity_and_content
+
+  let elem_of_vr = function `Valid e | `Removed e -> e
+
+  let contented_validity (validity:validity) content : 'a contented_validity =
+    match validity.removed with
+    | None -> { validity with content = `Valid content}
+    | Some _ -> { validity with content = `Removed content}
+
+  let introduced_in since = {valid = since ; removed = None; content = `Nothing }
+  let removed_since since ?replaced removal =
+    if since >= removal then
+      OpamConsole.error_and_exit `Internal_error
+        "An option can't be added in %s and removed in %s"
+        (OpamCLIVersion.to_string since)
+        (OpamCLIVersion.to_string removal);
+    { valid = since ; removed = Some (removal, replaced); content = `Nothing }
+  let cli_legacy = introduced_in cli2_0
+
+  let string_of_cli_option cli =
+    if cli = cli2_0 then
+      Printf.sprintf "set %s" (OpamConsole.colorise `bold "OPAMCLI=2.0")
+    else
+      Printf.sprintf "use --cli %s"
+        (OpamConsole.colorise `bold (OpamCLIVersion.to_string cli))
+
+  let update_doc_w_cli doc ~cli = function
+    | { valid = c ; removed = None; _} ->
+      if cli < c then
+        Printf.sprintf "(Since $(b,%s)) %s"
+          (OpamCLIVersion.to_string c) doc
+      else doc
+    | { removed = Some (since, instead); _} ->
+      if cli < since then doc else
+        Printf.sprintf "Removed since $(b,%s)%s"
```

Comment:
I think
```suggestion
        Printf.sprintf "Removed in $(b,%s)%s"
```
reads better, given that it's not the word deprecated. It's removed from the CLI, even if the support to do it hasn't been removed from opam.